third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/ConvertWarpSpecializeToLLVM.cpp

```diff
@@ -0,0 +1,427 @@
+    if (partition->getNumArguments()) {
+      auto captureType = LLVM::LLVMStructType::getLiteral(
+          b.getContext(), llvm::to_vector(partition->getArgumentTypes()));
+      Value capturePtr =
+          LLVM::getSharedMemoryBase(b.getLoc(), b, targetInfo, ws);
+      LLVM::LLVMPointerType ptrTy = ptr_ty(b.getContext(), 3);
+      for (auto [i, arg] :
+           llvm::zip(llvm::seq<int32_t>(partition->getNumArguments()),
+                     partition->getArguments())) {
+        Value ptr =
+            b.gep(ptrTy, captureType, capturePtr, ArrayRef<LLVM::GEPArg>{0, i});
+        // Each thread in the warp group needs a copy of the value.
+        Value value = b.load(arg.getType(), ptr);
+        arg.replaceAllUsesWith(value);
+      }
+      partition->front().eraseArguments([](auto) { return true; });
+    }
+
+    // The shared memory is only live for the entry into the region, so put
+    // another barrier here.
+    createBarrier(b, kSwitchLoopBarrierIdx, /*numThreads=*/std::nullopt,
+                  /*aligned=*/false);
+
+    // Rewrite all warp returns.
+    partition->walk([&](WarpReturnOp op) {
+      b.setInsertionPoint(op);
+      createBarrier(b, kSwitchLoopBarrierIdx, /*numThreads=*/std::nullopt,
+                    /*aligned=*/false);
+      b.replaceOpWithNewOp<LLVM::BrOp>(op, switchLoop);
+    });
+  }
+}
+
+static LogicalResult lowerWarpSpecialize(LLVM::LLVMFuncOp func,
+                                         const NVIDIA::TargetInfo &targetInfo) {
+  SmallVector<WarpSpecializeOp> wsOps;
+  func.walk([&](WarpSpecializeOp op) { wsOps.push_back(op); });
+  // Nothing to do. This kernel is not warp specialized.
+  if (wsOps.empty())
+    return success();
+
+  // Before lowering away `ttg.warp_specialize`, lower warp group barriers.
+  auto module = cast<ModuleOp>(func->getParentOp());
+  unsigned threadsPerWarp = TritonGPUDialect::getThreadsPerWarp(module);
+  unsigned defaultNumWarps = lookupNumWarps(func);
+  unsigned defaultWarpGroupSize = threadsPerWarp * defaultNumWarps;
+  if (failed(rewriteWarpGroupBarriers(func, wsOps, threadsPerWarp,
+                                      defaultWarpGroupSize)))
+    return failure();
+
+  MLIRContext *ctx = func.getContext();
+  TritonLLVMIRRewriter b(func.getLoc(), ctx);
+  Builder rewriter(ctx);
+
+  // Generate the function header.
+  Block *entry = &func.getBody().front();
+  SmallVector<Location> argLocs = llvm::to_vector(llvm::map_range(
+      func.getArguments(), [](BlockArgument arg) { return arg.getLoc(); }));
+  Block *header = b.createBlock(entry, func.getArgumentTypes(), argLocs);
+  Block *switchLoop = b.createBlock(entry);
+  b.setInsertionPointToStart(header);
+
+  // This is the absolute thread ID.
+  Value tid = b.create<NVVM::ThreadIdXOp>(i32_ty);
+  Value wid = b.udiv(tid, b.i32_val(threadsPerWarp));
+  // Tell PTXAS this value is warp-uniform.
+  wid = targetInfo.shuffleIdx(b, b.getLoc(), wid, 0);
```

Comment:
is this needed?

Reply:
I'm not sure, but I figured since this trick was used in a few other places that it might matter, especially for indicating that the branches are not divergent.

Reply:
ok, I guess that's a detail, at some point ptxas is supposed to be improved to figure it out based on reqid info, probably doesn't matter much and it's fine to leave for now